apis/v1alpha1/destination.go

```diff
@@ -143,6 +149,25 @@ func (dest *Destination) GetRef() *corev1.ObjectReference {
 	return nil
 }
 
+// NormalizeDeprecatedObjectReference copies the deprecated ObjectReference
+// fields to the Ref field and clears the deprecated fields. It uses the same
+// precedence rules as GetRef(). Use this to normalize code to use the Ref field
+// when deprecated fields may exist.
+// This method should be removed when the deprecated ObjectReference fields are
+// removed.
+func (dest *Destination) NormalizeDeprecatedObjectReference() {
```

Comment:
When do we expect this to be used? My understanding is that we _won't_ be automatically translating anything. Rather, we will just use deprecation warnings in Source's statuses.

Reply:
This will be used in source reconcilers. The reconciler needs to set the destination ref's namespace to the source's namespace, but since the ref can come from two places right now, that's more difficult than it should be.

I agree that we don't want to change the given object in Kubernetes storage. I'd expect reconcilers to never attempt to update the spec of the object (by using the status subresource) but I see your point that it may happen if we modify the ref in place.

Does it make sense to refactor this method to take a destination as argument and return a new normalized one? That way the original destination isn't modified.

Reply:
I see, this is intended to be used only inside a controller's reconcile loop and never have the altered destination be written out to storage.

If that is correct, then I don't think we should alter anything in place. I would probably change the method signature to:

```golang
func (dest Destination) NormalizeObjectReference() Destination
```

Then in controllers, wherever they want the normalized reference, they use:
```golang
source.sink.NormalizeObjectReference()
```

More generally, is this an improvement over just using Destionation's `GetRef()`?

Reply:
I updated the method to take a destination as argument and return a copied destination, similar to what you proposed. 

`GetRef()` doesn't work for this case because the reconciler must pass the entire Destination to `URIFromDestination` to get the uri reference resolution behavior.

Reply:
Can `URIFromDestination` use Destination's `GetRef`?

Reply:
That wouldn't help with this situation. If the reconciler always sets `Ref.Namespace`, `GetRef` will ignore the Deprecated fields. If the reconciler always sets `DeprecatedNamespace`, `GetRef` will ignore it if there's already a Ref. 

The behavior I want for the reconciler is to set Namespace in one place (ideally the `Ref` field). I'd rather each reconciler not have to care that there are two possible ways to specify a Ref:

```go
if dest.Ref != nil && dest.Ref.Name != "" {
  dest.Ref.Namespace = source.Namespace
} else if dest.DeprecatedName != "" {
  dest.Ref.DeprecatedNamespace = source.Namespace
}
```

But in the grand scheme of things it's not so bad, and it will only be around for one release, so actually now I'm thinking I should close this PR to avoid needing another pkg update in eventing.

Reply:
Got it, now I understand. Thank you for explaining it to me (repeatedly...). 

I now understand why this change is desirable, I'll leave it up to you if we want this, #810, or neither.